### PR TITLE
API: deprecate setting of .ordered directly (GH9347, GH9190)

### LINKED ISSUE #9148: Ordered vs. Unordered Categoricals
I was looking at supporting Categorical objects in statsmodels [1] recently and again now in seaborn [2]. I think it would be great to support both ordered and unordered Categorical variables (via a flag or something). IIRC, there was some discussion about this. Is it possible? From what I could tell all Categorical types are assumed to have some order.

[1] https://github.com/statsmodels/statsmodels/issues/2133
[2] https://github.com/mwaskom/seaborn/issues/361

### LINKED ISSUE #9190: Feature Request: categorical.reset_order
I was thinking about trying to add a `to_unordered` method, then I thought maybe a `reset_order` (or `reorder`?) with an optional `drop` keyword à la `reset_index` makes more sense. I didn't see if this was possible yet, so this is also a question. Is this possible via some other syntactic sugar? I might see if I can hack this together at some point unless someone beats me to it. My motivation for this is that I'm getting all ordered or unordered factors using `read_stata`. This could be "fixed" there by taking a list in addition to the boolean convert to ordered or whatever, but I think a method like this would be generally useful, plus I never peak at data before reading it.

### PR DESCRIPTION
**Pull request by jreback**
alternate to #9611 
closes #9347
closes #9190
closes #9148 

This implementes the alternate, and IMHO a nice compromise.

Groupby's will succeed whether ordered or not
ordering ops (sort/argsort/min/max) will still raise.

one though is that we could detect a non-reduction where the ordering matters (maybe we do this in another PR), and show a warning, e.g. imagine `df.groupby('A').head(2)` is technically 'wrong', while `df.groupby('A').sum()` would have no warning.

```
In [1]: df = DataFrame({ 'A' : Series(list('aabc')).astype('category'), 'B' : np.arange(4) })

In [2]: df.groupby('A').sum()
Out[2]: 
   B
A   
a  1
b  2
c  3

In [3]: df['A'].order()

TypeError: Categorical is not ordered for operation argsort
you can use .as_ordered() to change the Categorical to an ordered one
```

### CODE REVIEW

**Comment by jreback** [comment]
cc @JanSchulz
@jorisvandenbossche 
@shoyer

**Comment by JanSchulz** [comment]
LGTM... :-)

File: doc/source/categorical.rst [hunk 1]
```diff
@@ -281,18 +290,26 @@ raised.
         s.sort()
     except TypeError as e:
         print("TypeError: " + str(e))
-    s = Series(["a","b","c","a"], dtype="category") # ordered per default!
+    s = Series(["a","b","c","a"]).astype('category',ordered=True)
```
Comment (shoyer) [inline doc/source/categorical.rst]: nit: missing a space before `ordered=True`

File: doc/source/whatsnew/v0.16.0.txt [hunk 2]
```diff
@@ -365,6 +366,150 @@ API Changes
 - ``Series.describe`` for categorical data will now give counts and frequencies of 0, not ``NaN``, for unused categories (:issue:`9443`)
 
 
+Categorical Changes
+~~~~~~~~~~~~~~~~~~~
+
+.. _whatsnew_0160.api_breaking.categorical:
+
+In prior versions, ``Categoricals`` that had an unspecified ordering (meaning no ``ordered`` keyword was passed) were defaulted as ``ordered`` Categoricals. Going forward, the ``ordered`` keyword in the ``Categorical`` constructor will default to ``False``, ordering must now be explicit.
```
Comment (shoyer) [inline doc/source/whatsnew/v0.16.0.txt]: I would use a colon `:` before "ordering must now be explicit", not a comma.

File: doc/source/whatsnew/v0.16.0.txt [hunk 3]
```diff
@@ -365,6 +366,150 @@ API Changes
...
+   s = s.cat.as_ordered()
+   s
+   s.cat.ordered
+
+   # you can set in the constructor of the Categorical
+   s = Series(Categorical([0,1,2],ordered=True))
+   s
+   s.cat.ordered
+
+For ease of creation of series of categorical data, we have added the ability to pass keywords when calling ``.astype()``, these are passed directly to the constructor.
```
Comment (shoyer) [inline doc/source/whatsnew/v0.16.0.txt]: again, I would use a colon before the last clause or make it a new sentence

**Comment by shoyer** [comment]
Grammar nits aside, what happened to the idea of using `sort=None` in groupby? It seems a little bad to let `sort=True` pass silently for unordered categories.

> one though is that we could detect a non-reduction where the ordering matters (maybe we do this in another PR), and show a warning, e.g. imagine `df.groupby('A').head(2)` is technically 'wrong', while `df.groupby('A').sum()` would have no warning.

This probably reflects my ignorance of groupby internals, but to me, this looks more like a bug in `GroupBy.head` -- it doesn't respect `sort=True`:
```
In [17]: df = pd.DataFrame({'A': [2, 1, 0, 0], 'B': [0, 1, 2, 3]})

In [18]: df.groupby('A', sort=True).head(1)
Out[18]:
   A  B
0  2  0
1  1  1
2  0  2
```

Also, I didn't see a clear answer to @jorisvandenbossche's question about what `sort=None` means for the order of unordered categories. It looks like you're just using the existing categories in that case, which is fine by me -- I don't think this particularly matters. Though again, I think `sort=True` should raise in that case.

**Comment by jreback** [comment]
The sort only affects ``first`` and the like (my example above was wrong). ``.head()`` is a filter.
```
In [17]: df = DataFrame(np.random.randint(1, 10, (100, 2)),dtype='int64')

In [18]: df.groupby(df[1]).first()
Out[18]: 
   0
1   
1  7
2  6
3  9
4  6
5  7
6  7
7  1
8  9
9  6

In [19]: df.groupby(df[1],sort=False).first()
Out[19]: 
   0
1   
4  6
6  7
5  7
9  6
2  6
8  9
3  9
7  1
1  7
```

I think the groupby ``sort=None`` is not really an elegant soln and introduces some boilerplate.

**Comment by jorisvandenbossche** [comment]
To all, could you also react to my comment here (the previous PR):https://github.com/pydata/pandas/pull/9611#issuecomment-77928273

Why not allowing sorting? Because, sorting can also be seen as just the operation of putting the same categories together in the series. And this is something I potentially also want to do with unordered categories, without having to explicitely give an order to the categories.

**Comment by jreback** [comment]
@jorisvandenbossche 

ahh so, you want to all sort/argsort on even unordered categoricals, but disallow min/max. its fine by me as it makes less things break. its a bit non-strict, but as long as its clear that we order by categories I don't see any harm.

(but THAT is a change yes, not the order of appearance, but in order of the categoricals)?

**Comment by shoyer** [comment]
@jorisvandenbossche Hmm, interesting. I don't know if I have an educated opinion here. It does seem reasonable to me (especially since it's what R does).

**Comment by jorisvandenbossche** [comment]
@shoyer I also have to admit I don't really have an educated opinion. I was just triggered to look at it due to the discussion in the other PR, and then saw the behaviour of R. But the more I think about it, the more it seems natural to want to order a categorical even if is not strictly 'ordered'.

**Comment by jreback** [comment]
ok, so bottom line is that we allowe all operations on both ordered/unordered except for min/max?

**Comment by jorisvandenbossche** [comment]
cc @jseabold @mwaskom (the question is about allowing to sort an unordered categorical, see my comment above https://github.com/pydata/pandas/pull/9622#issuecomment-78166135 and here https://github.com/pydata/pandas/pull/9611#issuecomment-77928273) any opinions?

**Comment by jorisvandenbossche** [comment]
cc @njsmith

**Comment by jreback** [comment]
I updated [here]https://github.com/jreback/pandas/commit/7adad3be934f4b5c14582cfbe610e34d0e63b6ac)

**Comment by jreback** [comment]
we could always change later, but this needs to go in the rc.

File: pandas/core/panel.py
Comment (ac38618635-hue) [inline pandas/core/panel.py:1]: 5

**Review by ac38618635-hue — APPROVED** [review]

**Pull request merged**